File: packages/riff-openai-realtime/src/provider.ts
import type {
  ConnectRequest,
  ProviderCapabilities,
  ProviderEvent,
  RealtimeConnection,
  RealtimeProvider,
  SessionDefaults,
} from "@riff/core";
import { RiffProviderError, createEventHub } from "@riff/core";
import type { CredentialProvider, RealtimeTransport } from "./transport.ts";
import { decodeBase64 } from "./transport.ts";
import { CLIENT_EVENTS, SERVER_EVENTS, mapServerEvent } from "./events.ts";
import { biasingStyleFor, buildOpenAISession, buildVocabularyPatch } from "./session-config.ts";
import { connectWebSocket, type WebSocketLike } from "./websocket-transport.ts";
import { connectWebRTC, type RTCPeerConnectionLike } from "./webrtc-transport.ts";

export interface OpenAIRealtimeProviderOptions {
  credentials: CredentialProvider;
  /** `websocket` works everywhere; `webrtc` is better on a phone or in a browser. */
  transport?: "websocket" | "webrtc";
  /** Overrides the model in the bundle's session defaults. */
  model?: string;
  organization?: string;
  project?: string;
  /** Point at Azure OpenAI or a gateway. The event protocol is the same. */
  webSocketUrl?: string;
  callsUrl?: string;
  createWebSocket?: (url: string, protocols: string[]) => WebSocketLike;
  /** Required for the WebRTC transport, where the caller owns the peer connection and the mic. */
  webrtc?: {
    peerConnection: () => RTCPeerConnectionLike | Promise<RTCPeerConnectionLike>;
    tracks?: unknown[];
    onRemoteTrack?: (track: unknown, streams: unknown[]) => void;
  };
  /** Time to wait for `session.created` before giving up. */
  handshakeTimeoutMs?: number;
}

const HANDSHAKE_TIMEOUT_MS = 15_000;

/**
 * Riff on the OpenAI Realtime API.
 *
 * This class is the only place in the project that knows OpenAI's wire format. It implements the
 * provider interface and nothing more, which is what keeps the agent's behavior — the ledger, the
 * grounding check, the drafts — identical no matter what is generating the speech.
 */
export class OpenAIRealtimeProvider implements RealtimeProvider {
  readonly id = "openai-realtime";
  readonly #options: OpenAIRealtimeProviderOptions;

  constructor(options: OpenAIRealtimeProviderOptions) {
    this.#options = options;
    if (options.transport === "webrtc" && !options.webrtc) {
      throw new Error("the webrtc transport needs a webrtc.peerConnection factory");
    }
  }

  get capabilities(): ProviderCapabilities {
    const overWebRTC = this.#options.transport === "webrtc";
    return {
      speechToSpeech: true,
      bargeIn: true,
      semanticTurnDetection: true,
      vocabularyBiasing: "prompt",
      inputTranscription: true,
      functionCalling: true,
      audio: overWebRTC
        ? {
            input: { encoding: "opus", sampleRate: 48000, channels: 1 },
            output: { encoding: "opus", sampleRate: 48000, channels: 1 },
          }
        : {
            input: { encoding: "pcm_s16le", sampleRate: 24000, channels: 1 },
            output: { encoding: "pcm_s16le", sampleRate: 24000, channels: 1 },
          },
      maxSessionSeconds: 3600,
    };
  }

  /** How the configured transcription model wants vocabulary hints. */
  biasingStyle(session: SessionDefaults): "prompt" | "keywords" | "none" {
    return biasingStyleFor(session.transcription.preferred);
  }

  async connect(request: ConnectRequest): Promise<RealtimeConnection> {
    const model = this.#options.model ?? request.session.model.preferred;
    const hub = createEventHub();

    const transport = await this.#openTransport(model, request.signal);

    let sessionId = "";
    let negotiatedModel = model;
    let resolveHandshake: (() => void) | null = null;
    let rejectHandshake: ((error: Error) => void) | null = null;

    const handshake = new Promise<void>((resolve, reject) => {
      resolveHandshake = resolve;
      rejectHandshake = reject;
    });

    transport.onMessage((raw) => {
      const mapped = mapServerEvent(raw, decodeBase64);
      if (mapped.sessionId) {
        sessionId = mapped.sessionId;
        negotiatedModel = mapped.model ?? model;
        resolveHandshake?.();
        resolveHandshake = null;
      }
      // A failure during the handshake has to surface as a connect error, not a silent timeout.
      if (raw["type"] === SERVER_EVENTS.error && rejectHandshake) {
        const error = (raw["error"] ?? {}) as { message?: string; code?: string };
        rejectHandshake(new Error(error.message ?? "the realtime session was rejected"));
        rejectHandshake = null;
      }
      for (const event of mapped.events) hub.emit(event);
    });

    transport.onError((error) => {
      rejectHandshake?.(error);
      rejectHandshake = null;
      hub.emit({
        type: "error",
        error: { code: "transport_error", message: error.message, retryable: true, cause: error },
      });
    });

    transport.onClose((reason) => {
      rejectHandshake?.(new Error(reason ? `connection closed: ${reason}` : "connection closed"));
      rejectHandshake = null;
      hub.emit({ type: "closed", ...(reason ? { reason } : {}) });
    });

    try {
      await withTimeout(
        handshake,
        this.#options.handshakeTimeoutMs ?? HANDSHAKE_TIMEOUT_MS,
        "session.created",
        // Opening the transport is not the whole of connecting. Without this, an abort arriving
        // after the socket is open is ignored until the handshake timeout expires.
        request.signal,
      );
    } catch (error) {
      await transport.close("handshake failed");
      throw new RiffProviderError("handshake_failed", (error as Error).message, { retryable: true, cause: error });
    }

    transport.send({
      type: CLIENT_EVENTS.sessionUpdate,
      session: buildOpenAISession({
        session: request.session,
        instructions: request.instructions,
        tools: request.tools,
        ...(request.vocabulary ? { vocabulary: request.vocabulary } : {}),
        model,
      }),
    });

    return this.#createConnection(transport, hub, request.session, sessionId, negotiatedModel);
  }

  async #openTransport(model: string, signal?: AbortSignal): Promise<RealtimeTransport> {
    if (this.#options.transport === "webrtc") {
      const webrtc = this.#options.webrtc!;
      return connectWebRTC({
        model,
        credentials: this.#options.credentials,
        peerConnection: await webrtc.peerConnection(),
        ...(this.#options.callsUrl ? { url: this.#options.callsUrl } : {}),
        ...(webrtc.tracks ? { tracks: webrtc.tracks } : {}),
        ...(webrtc.onRemoteTrack ? { onRemoteTrack: webrtc.onRemoteTrack } : {}),
        ...(signal ? { signal } : {}),
      });
    }

    return connectWebSocket({
      model,
      credentials: this.#options.credentials,
      ...(this.#options.webSocketUrl ? { url: this.#options.webSocketUrl } : {}),
      ...(this.#options.organization ? { organization: this.#options.organization } : {}),
      ...(this.#options.project ? { project: this.#options.project } : {}),
      ...(this.#options.createWebSocket ? { createWebSocket: this.#options.createWebSocket } : {}),
      ...(signal ? { signal } : {}),
    });
  }

  #createConnection(
    transport: RealtimeTransport,
    hub: { emit: (event: ProviderEvent) => void; on: (listener: (event: ProviderEvent) => void) => () => void },
    session: SessionDefaults,
    sessionId: string,
    model: string,
  ): RealtimeConnection {
    return {
      sessionId,
      model,

      sendAudio(chunk) {
        transport.sendAudio(chunk);
      },

      commitAudio() {
        transport.send({ type: CLIENT_EVENTS.commitAudio });
      },

      sendText(text, options) {
        transport.send({
          type: CLIENT_EVENTS.createItem,
          item: { type: "message", role: "user", content: [{ type: "input_text", text }] },
        });
        if (options?.respond !== false) transport.send({ type: CLIENT_EVENTS.createResponse });
      },

      respondToTool(callId, resultJson) {
        transport.send({
          type: CLIENT_EVENTS.createItem,
          item: { type: "function_call_output", call_id: callId, output: resultJson },
        });
      },

      requestResponse() {
        transport.send({ type: CLIENT_EVENTS.createResponse });
      },

      cancelResponse() {
        transport.send({ type: CLIENT_EVENTS.cancelResponse });
        // On WebRTC the already-buffered audio keeps playing unless it is explicitly dropped, so
        // the agent would talk over someone who just interrupted it.
        if (transport.kind === "webrtc") transport.send({ type: "output_audio_buffer.clear" });
      },

      updateSession(patch) {
        if (patch.vocabulary) {
          const vocabularyPatch = buildVocabularyPatch(session, patch.vocabulary);
          if (vocabularyPatch) transport.send({ type: CLIENT_EVENTS.sessionUpdate, session: vocabularyPatch });
        }

        const { vocabulary: _vocabulary, ...rest } = patch;
        if (Object.keys(rest).length === 0) return;

        transport.send({
          type: CLIENT_EVENTS.sessionUpdate,
          session: buildOpenAISession({
            session: { ...session, ...rest } as SessionDefaults,
            instructions: "",
            tools: [],
          }),
        });
      },

      on(listener) {
        return hub.on(listener);
      },

      async close(reason) {
        await transport.close(reason);
      },
    };
  }
}

function withTimeout<T>(promise: Promise<T>, ms: number, what: string, signal?: AbortSignal): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    if (signal?.aborted) return reject(new Error("connection aborted"));

    const timer = setTimeout(() => reject(new Error(`timed out after ${ms}ms waiting for ${what}`)), ms);
    timer.unref?.();

    const onAbort = () => {
      clearTimeout(timer);
      reject(new Error("connection aborted"));
    };
    signal?.addEventListener("abort", onAbort, { once: true });

    const settle = () => {
      clearTimeout(timer);
      signal?.removeEventListener("abort", onAbort);
    };

    promise.then(
      (value) => {
        settle();
        resolve(value);
      },
      (error) => {
        settle();
        reject(error);
      },
    );
  });
}

File: rust/riff-openai-realtime/src/provider.rs
//! Riff on the OpenAI Realtime API.
//!
//! This is the only place in the Rust binding that knows OpenAI's wire format. It implements the
//! provider interface and nothing more, which is what keeps the agent's behavior — the ledger, the
//! grounding check, the drafts — identical no matter what is generating the speech.

use std::collections::VecDeque;
use std::sync::{Arc, Mutex};

use riff_core::{
    AudioFormat, BoxFuture, Clock, ConnectRequest, Either, Json, ProviderCapabilities,
    ProviderEvent, ProviderFault, RealtimeConnection, RealtimeProvider, SessionDefaults,
    SystemClock, json_object, race,
};

use crate::events::{client_events, map_server_event};
use crate::session_config::{
    BiasingStyle, BuildSessionOptions, biasing_style_for, build_openai_session,
    build_vocabulary_patch,
};
use crate::transport::{
    CredentialProvider, RealtimeTransport, TransportFactory, TransportKind, TransportMessage,
    TransportRequest, encode_base64, endpoint_with_model,
};

/// The default WebSocket endpoint.
pub const DEFAULT_WEBSOCKET_URL: &str = "wss://api.openai.com/v1/realtime";

/// How long to wait for `session.created` before giving up.
const HANDSHAKE_TIMEOUT_MS: u64 = 15_000;

/// How the provider is put together.
pub struct OpenAIRealtimeOptions {
    /// What supplies the token. Use [`crate::ApiKeyCredentials`] only on a server.
    pub credentials: Arc<dyn CredentialProvider>,
    /// What opens the socket.
    pub transport: Arc<dyn TransportFactory>,
    /// Time and delays, which bound the handshake.
    pub clock: Arc<dyn Clock>,
    /// Overrides the model in the bundle's session defaults.
    pub model: Option<String>,
    /// The OpenAI organization, when the account has more than one.
    pub organization: Option<String>,
    /// The OpenAI project.
    pub project: Option<String>,
    /// Point at Azure OpenAI or a gateway. The event protocol is the same.
    pub url: Option<String>,
    /// How long to wait for `session.created`.
    pub handshake_timeout_ms: u64,
}

impl OpenAIRealtimeOptions {
    /// The usual configuration: a credential source and a transport, everything else defaulted.
    pub fn new(
        credentials: Arc<dyn CredentialProvider>,
        transport: Arc<dyn TransportFactory>,
    ) -> Self {
        Self {
            credentials,
            transport,
            clock: Arc::new(SystemClock::new()),
            model: None,
            organization: None,
            project: None,
            url: None,
            handshake_timeout_ms: HANDSHAKE_TIMEOUT_MS,
        }
    }
}

/// The provider.
pub struct OpenAIRealtimeProvider {
    options: OpenAIRealtimeOptions,
}

impl OpenAIRealtimeProvider {
    /// Builds the provider.
    pub fn new(options: OpenAIRealtimeOptions) -> Self {
        Self { options }
    }

    /// How the configured transcription model wants vocabulary hints.
    pub fn biasing_style(&self, session: &SessionDefaults) -> BiasingStyle {
        biasing_style_for(&session.transcription.preferred)
    }
}

impl RealtimeProvider for OpenAIRealtimeProvider {
    fn id(&self) -> &str {
        "openai-realtime"
    }

    fn capabilities(&self) -> ProviderCapabilities {
        let pcm = AudioFormat {
            encoding: "pcm_s16le".to_owned(),
            sample_rate: 24_000,
            channels: 1,
        };
        ProviderCapabilities {
            speech_to_speech: true,
            barge_in: true,
            semantic_turn_detection: true,
            vocabulary_biasing: "prompt".to_owned(),
            input_transcription: true,
            function_calling: true,
            input_audio: pcm.clone(),
            output_audio: pcm,
            max_session_seconds: Some(3600),
        }
    }

    fn connect(
        &self,
        request: ConnectRequest,
    ) -> BoxFuture<'_, Result<Arc<dyn RealtimeConnection>, ProviderFault>> {
        Box::pin(async move {
            let model = self
                .options
                .model
                .clone()
                .unwrap_or_else(|| request.session.model.preferred.clone());
            let credential = self.options.credentials.credential().await?;

            let transport = self
                .options
                .transport
                .open(TransportRequest {
                    url: endpoint_with_model(
                        self.options.url.as_deref().unwrap_or(DEFAULT_WEBSOCKET_URL),
                        &model,
                    ),
                    credential,
                    organization: self.options.organization.clone(),
                    project: self.options.project.clone(),
                })
                .await?;

            let handshake = await_handshake(
                transport.as_ref(),
                self.options.clock.as_ref(),
                self.options.handshake_timeout_ms,
            )
            .await;

            let handshake = match handshake {
                Ok(handshake) => handshake,
                Err(fault) => {
                    transport.close(Some("handshake failed".to_owned())).await;
                    return Err(fault);
                }
            };

            transport.send(&Json::Object(json_object! {
                "type" => client_events::SESSION_UPDATE,
                "session" => build_openai_session(BuildSessionOptions {
                    session: &request.session,
                    instructions: &request.instructions,
                    tools: &request.tools,
                    vocabulary: &request.vocabulary,
                    model: Some(&model),
                }),
            }));

            Ok(Arc::new(OpenAIConnection {
                transport,
                session: request.session,
                session_id: handshake.session_id,
                model: handshake.model,
                pending: Mutex::new(handshake.pending),
            }) as Arc<dyn RealtimeConnection>)
        })
    }
}

struct Handshake {
    session_id: String,
    model: String,
    /// Events that arrived during the handshake, so `session.created` still reaches the session.
    pending: VecDeque<ProviderEvent>,
}

/// Waits for `session.created`, or for something that says it will never come.
///
/// A failure during the handshake has to surface as a connect error rather than as a silent timeout,
/// so a rejected session configuration is reported as what it is.
async fn await_handshake(
    transport: &dyn RealtimeTransport,
    clock: &dyn Clock,
    timeout_ms: u64,
) -> Result<Handshake, ProviderFault> {
    let mut pending = VecDeque::new();
    // Created once and re-polled across iterations, and raced ahead of the message stream. A fresh
    // delay per message would bound only the gap between messages; a delay raced second would never
    // be polled at all while an endpoint keeps a keepalive ready. Either way `connect` would hang.
    let mut deadline = (timeout_ms > 0).then(|| clock.sleep(timeout_ms));

    loop {
        let message = match deadline.as_mut() {
            None => transport.next_message().await,
            Some(deadline) => match race(deadline, transport.next_message()).await {
                Either::Left(()) => {
                    return Err(ProviderFault::retryable(
                        "handshake_failed",
                        format!("timed out after {timeout_ms}ms waiting for session.created"),
                    ));
                }
                Either::Right(message) => message,
            },
        };

        match message {
            Some(TransportMessage::Event(raw)) => {
                let mapped = map_server_event(&raw);
                let identity = mapped
                    .session_id
                    .clone()
                    .map(|session_id| (session_id, mapped.model.clone().unwrap_or_default()));
                let failure = mapped.events.iter().find_map(|event| match event {
                    ProviderEvent::Failed { fault } => Some(fault.clone()),
                    _ => None,
                });
                pending.extend(mapped.events);

                if let Some(fault) = failure {
                    return Err(ProviderFault::retryable("handshake_failed", fault.message));
                }
                if let Some((session_id, model)) = identity {
                    return Ok(Handshake {
                        session_id,
                        model,
                        pending,
                    });
                }
            }
            Some(TransportMessage::Failed(fault)) => {
                return Err(ProviderFault::retryable("handshake_failed", fault.message));
            }
            Some(TransportMessage::Closed(reason)) => {
                return Err(closed_during_handshake(reason));
            }
            None => return Err(closed_during_handshake(None)),
        }
    }
}

fn closed_during_handshake(reason: Option<String>) -> ProviderFault {
    ProviderFault::retryable(
        "handshake_failed",
        match reason {
            Some(reason) => format!("connection closed: {reason}"),
            None => "connection closed".to_owned(),
        },
    )
}

struct OpenAIConnection {
    transport: Arc<dyn RealtimeTransport>,
    session: SessionDefaults,
    session_id: String,
    model: String,
    pending: Mutex<VecDeque<ProviderEvent>>,
}

impl OpenAIConnection {
    fn take_pending(&self) -> Option<ProviderEvent> {
        self.pending
            .lock()
            .unwrap_or_else(|error| error.into_inner())
            .pop_front()
    }

    fn push_pending<I: IntoIterator<Item = ProviderEvent>>(&self, events: I) {
        self.pending
            .lock()
            .unwrap_or_else(|error| error.into_inner())
            .extend(events);
    }

    fn create_item(&self, item: Json) {
        self.transport.send(&Json::Object(json_object! {
            "type" => client_events::CREATE_ITEM,
            "item" => item,
        }));
    }
}

impl RealtimeConnection for OpenAIConnection {
    fn session_id(&self) -> String {
        self.session_id.clone()
    }

    fn model(&self) -> String {
        self.model.clone()
    }

    fn next_event(&self) -> BoxFuture<'_, Option<ProviderEvent>> {
        Box::pin(async move {
            loop {
                if let Some(event) = self.take_pending() {
                    return Some(event);
                }
                match self.transport.next_message().await {
                    Some(TransportMessage::Event(raw)) => {
                        self.push_pending(map_server_event(&raw).events);
                    }
                    Some(TransportMessage::Failed(fault)) => {
                        return Some(ProviderEvent::Failed { fault });
                    }
                    Some(TransportMessage::Closed(reason)) => {
                        return Some(ProviderEvent::Closed { reason });
                    }
                    None => return None,
                }
            }
        })
    }

    fn send_audio(&self, chunk: &[u8]) {
        if self.transport.kind() == TransportKind::WebRtc {
            // Audio travels on the negotiated media track, which the host attached itself.
            self.transport.send_audio(chunk);
            return;
        }
        self.transport.send(&Json::Object(json_object! {
            "type" => client_events::APPEND_AUDIO,
            "audio" => encode_base64(chunk),
        }));
    }

    fn commit_audio(&self) {
        self.transport.send(&Json::Object(
            json_object! { "type" => client_events::COMMIT_AUDIO },
        ));
    }

    fn send_text(&self, text: &str, respond: bool) {
        self.create_item(Json::Object(json_object! {
            "type" => "message",
            "role" => "user",
            "content" => vec![Json::Object(json_object! {
                "type" => "input_text",
                "text" => text,
            })],
        }));
        if respond {
            self.request_response();
        }
    }

    fn respond_to_tool(&self, call_id: &str, result_json: &str) {
        self.create_item(Json::Object(json_object! {
            "type" => "function_call_output",
            "call_id" => call_id,
            "output" => result_json,
        }));
    }

    fn request_response(&self) {
        self.transport.send(&Json::Object(
            json_object! { "type" => client_events::CREATE_RESPONSE },
        ));
    }

    fn cancel_response(&self) {
        self.transport.send(&Json::Object(
            json_object! { "type" => client_events::CANCEL_RESPONSE },
        ));
        // On WebRTC the already-buffered audio keeps playing unless it is explicitly dropped, so the
        // agent would talk over someone who just interrupted it.
        if self.transport.kind() == TransportKind::WebRtc {
            self.transport.send(&Json::Object(
                json_object! { "type" => client_events::CLEAR_OUTPUT_AUDIO },
            ));
        }
    }

    fn update_vocabulary(&self, vocabulary: &[String]) {
        if let Some(patch) = build_vocabulary_patch(&self.session, vocabulary) {
            self.transport.send(&Json::Object(json_object! {
                "type" => client_events::SESSION_UPDATE,
                "session" => patch,
            }));
        }
    }

    fn close(&self, reason: Option<String>) -> BoxFuture<'_, ()> {
        self.transport.close(reason)
    }
}
